Should we cap the required_python version in v1?

### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))


### Description

It was asked why we should cap the version in Pydantic V1 to an upper bound of 3.12: https://github.com/pydantic/pydantic/issues/9622#issuecomment-2165288361

My personal response was:

> Its "explicit support" vs "implied support"
>
> It may work in most usecases in 3.13. It may even pass the full test suite, though I have strong doubts on the cythonized builds working. In the end, just because it works doesn't mean its supported.
>
> It may be nontrivial to add full 3.13 support or maybe they don't want to be on the hook for dealing with another full life cycle of a release because there may be compatibility issues between bugfix versions like we just had with 3.12.4 and they need to focus developer time elsewhere, like pydantic v2 or a new product, instead of on a pydantic 1.10.x thats in maintenance mode.

To add additional detail to the "new version support", the challenges of ensuring that pydantic work on a new version are, at a minimum:
* leveraged private functions in the stdlib haven't changed within python. Private functions are under no contract to maintain compatibility
* dependencies used by pydantic for runtime and testing are also available or buildable for the target version
* For cythonized builds: cython has support for the target version (Cython 0.x will almost certainly not see Python 3.13 support.)
  * Pure python based builds (sdist, wheels) may still be released without the cython requirement

After responding to the original question, however, I decided to dig into this a bit and I think the unfortunate reality is that the Python packaging environment is _not at all friendly to upper bound Python versions_.

For some "light reading", see:
* https://discuss.python.org/t/requires-python-upper-limits/12663
* https://github.com/scipy/scipy/issues/14738#issuecomment-997236377
* https://github.com/numpy/numpy/issues/24810
* https://github.com/scipy/scipy/pull/16553
* https://github.com/python-poetry/poetry/issues/6884
* https://github.com/python-poetry/poetry/issues/6921
* https://github.com/numba/numba/blob/0994f97c33a19d7684471cc98b8c36573762204b/setup.py#L44-L47

So, while our intention may be to cap the upper bounds, puzzle solvers may just **fall back to an older version** and install that since it was released without a cap, which may be undesired. It may also have unintended consequences for dependency lockfiles for certain package managers (like poetry).

Until the semantics are worked out with the packaging group, it may make sense to undo the cap I added in the v1 branch and instead add a build time check for the python version and issue a `RuntimeError` like numba has done. This would allow attempts at using the sdist install on an unsupported version to throw an error, though it may be good to cover __init__ for the pure python wheel too maybe... With the caveat that the copied sources in the v2 branch will need have a fixup via `update_v1.sh` to handle this by dropping the runtime error.

The last thing I want to do is actually cause more headaches for packages that use Pydantic.

### Example Code

_No response_

### Python, Pydantic & OS Version

```Text
pydantic 1.10.16+, any
```


### Affected Components

- [X] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

The release is scheduled in October. I expect it will take time for all dependency authors to update their packages unless they start testing on 3.13.0b now. For example, I think I ran into issues building greenlet and orjson in a 3.13 environment, required for building the documentation.

Right now, the failures on 3.13.0b2 are:

A half expected Cython issue:
```
FAILED tests/test_edge_cases.py::test_cython_function_untouched - distutils.errors.CompileError: command '/usr/bin/gcc' failed with exit code 1
tests/test_edge_cases.py:1877 test_cython_function_untouched


Compiling /home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.pyx because it changed.
[1/1] Cythonizing /home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.pyx
-------------------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------------------
warning: /home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.pyx:16:4: Unreachable code
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:840:1: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]
  840 | static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u) {
      | ^~~~~~
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c: In function ‘__Pyx_Py_UNICODE_strlen’:
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:841:5: warning: ‘Py_UNICODE’ is deprecated [-Wdeprecated-declarations]
  841 |     const Py_UNICODE *u_end = u;
      |     ^~~~~
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c: In function ‘__pyx_pf_55_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7___invoke’:
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:1059:38: warning: implicit declaration of function ‘_PyDict_SetItem_KnownHash’; did you mean ‘_PyDict_GetItem_KnownHash’? [-Wimplicit-function-declaration]
 1059 |     (likely(PyDict_CheckExact(ns)) ? _PyDict_SetItem_KnownHash(ns, name, value, ((PyASCIIObject *) name)->hash) : PyObject_SetItem(ns, name, value))
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:1487:7: note: in expansion of macro ‘__Pyx_SetNameInClass’
 1487 |   if (__Pyx_SetNameInClass(__pyx_t_3, __pyx_n_s_a, __pyx_float_0_0) < 0) __PYX_ERR(0, 9, __pyx_L1_error)
      |       ^~~~~~~~~~~~~~~~~~~~
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c: In function ‘__Pyx_PyInt_As_long’:
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:3385:27: error: too few arguments to function ‘_PyLong_AsByteArray’
 3385 |                 int ret = _PyLong_AsByteArray((PyLongObject *)v,
      |                           ^~~~~~~~~~~~~~~~~~~
In file included from /home/vfazio/.pyenv/versions/3.13.0b2/include/python3.13/longobject.h:107,
                 from /home/vfazio/.pyenv/versions/3.13.0b2/include/python3.13/Python.h:72,
                 from /home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:18:
/home/vfazio/.pyenv/versions/3.13.0b2/include/python3.13/cpython/longobject.h:108:17: note: declared here
  108 | PyAPI_FUNC(int) _PyLong_AsByteArray(PyLongObject* v,
      |                 ^~~~~~~~~~~~~~~~~~~
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c: In function ‘__Pyx_PyInt_As_int’:
/home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:3581:27: error: too few arguments to function ‘_PyLong_AsByteArray’
 3581 |                 int ret = _PyLong_AsByteArray((PyLongObject *)v,
      |                           ^~~~~~~~~~~~~~~~~~~
In file included from /home/vfazio/.pyenv/versions/3.13.0b2/include/python3.13/longobject.h:107,
                 from /home/vfazio/.pyenv/versions/3.13.0b2/include/python3.13/Python.h:72,
                 from /home/vfazio/.cython/inline/_cython_inline_0f57de3e5e4c42b1264f1bac2c6226ce4ba714c7.c:18:
/home/vfazio/.pyenv/versions/3.13.0b2/include/python3.13/cpython/longobject.h:108:17: note: declared here
  108 | PyAPI_FUNC(int) _PyLong_AsByteArray(PyLongObject* v,
      |                 ^~~~~~~~~~~~~~~~~~~


```

Private attribute errors:
```
FAILED tests/test_private_attributes.py::test_private_attribute_annotation - AssertionError: assert {'__firstline...attributes__'} == {'__foo__'}
FAILED tests/test_private_attributes.py::test_underscore_attrs_are_private - AssertionError: assert {'__firstline...attributes__'} == {'__foo__'}


    def test_private_attribute_annotation():
        class Model(BaseModel):
            """The best model"""
    
            __foo__: str
    
            class Config:
                underscore_attrs_are_private = True
    
>       assert Model.__slots__ == {'__foo__'}
E       AssertionError: assert {'__firstline...attributes__'} == {'__foo__'}
E         Extra items in the left set:
E         '__static_attributes__'
E         '__firstlineno__'
E         Use -v to get more diff

tests/test_private_attributes.py:66: AssertionError

 tests/test_private_attributes.py ⨯                                                                                                                                                                                                             55% █████▌    

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_underscore_attrs_are_private ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    def test_underscore_attrs_are_private():
        class Model(BaseModel):
            __foo__: str = 'abc'
            __bar__: ClassVar[str] = 'cba'
    
            class Config:
                underscore_attrs_are_private = True
    
>       assert Model.__slots__ == {'__foo__'}
E       AssertionError: assert {'__firstline...attributes__'} == {'__foo__'}
E         Extra items in the left set:
E         '__static_attributes__'
E         '__firstlineno__'
E         Use -v to get more diff

tests/test_private_attributes.py:101: AssertionError

```

And a number of these:

```
FAILED tests/test_validators.py::test_make_generic_validator[fields0-_v_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields1-_v_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields2-_v_,_field_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields3-_v_,_config_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields4-_v_,_values_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields5-_v_,_field_,_config_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields6-_v_,_field_,_values_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields7-_v_,_config_,_values_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields8-_v_,_field_,_values_,_config_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields9-_cls_,_v_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields10-_cls_,_v_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields11-_cls_,_v_,_field_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields12-_cls_,_v_,_config_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields13-_cls_,_v_,_values_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields14-_cls_,_v_,_field_,_config_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields15-_cls_,_v_,_field_,_values_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields16-_cls_,_v_,_config_,_values_] - KeyError: 'testing_function'
FAILED tests/test_validators.py::test_make_generic_validator[fields17-_cls_,_v_,_field_,_values_,_config_] - KeyError: 'testing_function'

 tests/test_validators.py ⨯                                                                                                                                                                                                                     96% █████████▋

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_make_generic_validator[fields17-_cls_,_v_,_field_,_values_,_config_] ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

fields = ['cls', 'val', 'field', 'values', 'config'], result = '_cls_,_v_,_field_,_values_,_config_'

    @pytest.mark.parametrize(
        'fields,result',
        [
            (['val'], '_v_'),
            (['foobar'], '_v_'),
            (['val', 'field'], '_v_,_field_'),
            (['val', 'config'], '_v_,_config_'),
            (['val', 'values'], '_v_,_values_'),
            (['val', 'field', 'config'], '_v_,_field_,_config_'),
            (['val', 'field', 'values'], '_v_,_field_,_values_'),
            (['val', 'config', 'values'], '_v_,_config_,_values_'),
            (['val', 'field', 'values', 'config'], '_v_,_field_,_values_,_config_'),
            (['cls', 'val'], '_cls_,_v_'),
            (['cls', 'foobar'], '_cls_,_v_'),
            (['cls', 'val', 'field'], '_cls_,_v_,_field_'),
            (['cls', 'val', 'config'], '_cls_,_v_,_config_'),
            (['cls', 'val', 'values'], '_cls_,_v_,_values_'),
            (['cls', 'val', 'field', 'config'], '_cls_,_v_,_field_,_config_'),
            (['cls', 'val', 'field', 'values'], '_cls_,_v_,_field_,_values_'),
            (['cls', 'val', 'config', 'values'], '_cls_,_v_,_config_,_values_'),
            (['cls', 'val', 'field', 'values', 'config'], '_cls_,_v_,_field_,_values_,_config_'),
        ],
    )
    def test_make_generic_validator(fields, result):
        exec(f"""def testing_function({', '.join(fields)}): return {' + "," + '.join(fields)}""")
>       func = locals()['testing_function']
E       KeyError: 'testing_function'

tests/test_validators.py:751: KeyError

```

However, regardless of the timeline for the release of 3.13 is whether pydantic wants to extend the life of v1 through 3.13. They may be signing up for more than they really want by doing that.

The two questions that need answering are:

1. What's the EOL and last major python release for the v1 API
2. What's the proper way to advertise and enforce that cap (`requires_python` or some other mechanism)

--- a/setup.py
+++ b/setup.py
@@ -134,7 +134,7 @@ setup(
     license='MIT',
     packages=['pydantic'],
     package_data={'pydantic': ['py.typed']},
-    python_requires='>=3.7, <3.13',
+    python_requires='>=3.7',
     zip_safe=False,  # https://mypy.readthedocs.io/en/latest/installed_packages.html
     install_requires=[
         'typing-extensions>=4.2.0'

